dsmr/p2p.go

```diff
@@ -0,0 +1,237 @@
+			Message: err.Error(),
+		}
+	}
+
+	chunkID, err := ids.ToID(request.ChunkId)
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: err.Error(),
+		}
+	}
+
+	// TODO check chunk status?
+	chunkBytes, available, err := g.storage.GetChunkBytes(request.Expiry, chunkID)
+	if err != nil && errors.Is(err, database.ErrNotFound) {
+		return nil, ErrChunkNotAvailable
+	}
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: err.Error(),
+		}
+	}
+
+	if !available {
+		return nil, ErrChunkNotAvailable
+	}
+
+	chunk, err := ParseChunk[T](chunkBytes)
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: err.Error(),
+		}
+	}
+
+	txs := make([]*dsmr.Transaction, 0, len(chunk.Txs))
+	for _, tx := range chunk.Txs {
+		packer := &wrappers.Packer{MaxSize: consts.NetworkSizeLimit}
+		if err := codec.LinearCodec.MarshalInto(tx, packer); err != nil {
+			panic(err)
+		}
+
+		txs = append(txs, &dsmr.Transaction{Bytes: packer.Bytes})
+	}
+
+	response := &dsmr.GetChunkResponse{
+		Chunk: &dsmr.Chunk{
+			Producer:     chunk.Producer[:],
+			Expiry:       chunk.Expiry,
+			Beneficiary:  chunk.Beneficiary[:],
+			Transactions: txs,
+			Signer:       chunk.Signer[:],
+			Signature:    chunk.Signature[:],
+		},
+	}
+
+	responseBytes, err := proto.Marshal(response)
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    common.ErrUndefined.Code,
+			Message: err.Error(),
+		}
+	}
+
+	return responseBytes, nil
+}
+
+// TODO rename, this is a storage type
+type ChunkSignature struct {
+	ChunkID    ids.ID
+	ProducerID ids.NodeID
+	Expiry     time.Time
+	Signer     [bls.PublicKeyLen]byte
+	Signature  [bls.SignatureLen]byte
+}
+
+type GetChunkSignatureHandler[T Tx] struct {
+	sk       *bls.SecretKey
+	verifier Verifier[T]
+	storage  *chunkStorage[T]
+}
+
+func (*GetChunkSignatureHandler[T]) AppGossip(context.Context, ids.NodeID, []byte) {}
+
+func (g *GetChunkSignatureHandler[T]) AppRequest(
+	_ context.Context,
+	_ ids.NodeID,
+	_ time.Time,
+	requestBytes []byte,
+) ([]byte, *common.AppError) {
+	request := &dsmr.GetChunkSignatureRequest{}
+	if err := proto.Unmarshal(requestBytes, request); err != nil {
+		return nil, &common.AppError{
+			Code:    p2p.ErrUnexpected.Code,
+			Message: err.Error(),
+		}
+	}
+
+	chunk, err := newChunkFromProto[T](request.Chunk)
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    p2p.ErrUnexpected.Code,
+			Message: err.Error(),
+		}
+	}
+
+	if err := g.verifier.Verify(chunk); err != nil {
+		return nil, ErrInvalidChunk
+	}
+
+	_, accepted, err := g.storage.GetChunkBytes(chunk.Expiry, chunk.id)
+	if err != nil && !errors.Is(err, database.ErrNotFound) {
+		return nil, &common.AppError{
+			Code:    p2p.ErrUnexpected.Code,
+			Message: err.Error(),
+		}
+	}
+
+	if accepted {
+		// Don't sign a chunk that is already marked as accepted
+		return nil, ErrDuplicateChunk
+	}
```

Comment:
The spec explicitly mentioned this so I included it. I don't think this is actually needed because you'd be checking for this as well during block verification...